File: chapter1/008/main.py
def main():
    matrix = [
        [1,1,1,1],
        [2,2,0,2],
        [3,0,3,3],
        [4,4,4,4]
    ]
    print_matrix(matrix)

    matrix = [
        [1,1,1],
        [2,0,2],
        [3,3,3]
    ]
    print_matrix(matrix)

    matrix = [
        [1,1],
        [2,0]
    ]
    print_matrix(matrix)

    matrix = [
        [1]
    ]
    print_matrix(matrix)

    matrix = [
        []
    ]
    print_matrix(matrix)

def zero(matrix):
    if len(matrix) == 0:
        return

    num_lines = len(matrix)
    num_cols = len(matrix[0])

    lines = [False] * num_lines
    columns = [False] * num_cols

    for i in range(num_lines):
        for j in range(num_cols):
            if matrix[i][j] == 0:
                lines[i] = True
                columns[j] = True
    
    for i, should_zero in enumerate(lines):
        if should_zero:
            zeroLine(matrix, i)
    
    for j, should_zero in enumerate(columns):
        if should_zero:
            zeroColumn(matrix, j)

def zeroLine(matrix, i):
    for j in range(len(matrix[0])):
        matrix[i][j] = 0

def zeroColumn(matrix, j):
    for i in range(len(matrix)):
       matrix[i][j] = 0

def print_matrix(matrix):
    for l in matrix:
        print(l)
    print("--------------------")
    zero(matrix)
    for l in matrix:
        print(l)
    print("===================+")

main()

File: chapter1/008/main.go
package main

import (
	"fmt"
)

func main() {
	matrix := [][]int{
		{1, 2, 3, 0, 0, 0},
		{4, 5, 6, 0, 0, 0},
		{7, 8, 9, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
		{10, 11, 12, 0, 0, 0},
	}
	printMatrix(matrix)

	matrix = [][]int{
		{1, 2, 3, 1},
		{4, 5, 6, 1},
		{0, 8, 0, 1},
		{10, 11, 12, 1},
	}
	printMatrix(matrix)

	matrix = [][]int{
		{1, 2, 3},
		{4, 0, 6},
		{7, 8, 9},
	}
	printMatrix(matrix)

	matrix = [][]int{
		{1, 2},
		{3, 4},
	}
	printMatrix(matrix)

	matrix = [][]int{
		{1},
	}
	printMatrix(matrix)
}

func zero(matrix *[][]int) {
	m := *matrix

	lines := make([]bool, len(m))
	columns := make([]bool, len(m[0]))

	for i, l := range m {
		for j, e := range l {
			if e == 0 {
				lines[i] = true
				columns[j] = true
			}
		}
	}

	zeroLines(matrix, lines...)
	zeroColumns(matrix, columns...)
}

func zeroLines(matrix *[][]int, lines ...bool) {
	m := *matrix
	for l, shouldZero := range lines {
		if shouldZero {
			for i := 0; i < len(m[0]); i++ {
				m[l][i] = 0
			}
		}
	}
}

func zeroColumns(matrix *[][]int, columns ...bool) {
	m := *matrix
	for c, shouldZero := range columns {
		if shouldZero {
			for i := 0; i < len(m); i++ {
				m[i][c] = 0
			}
		}
	}
}

func printMatrix(matrix [][]int) {
	for _, l := range matrix {
		for _, e := range l {
			fmt.Printf("%d\t", e)
		}
		fmt.Println()
	}
	zero(&matrix)
	fmt.Println("---------------------------------")
	for _, l := range matrix {
		for _, e := range l {
			fmt.Printf("%d\t", e)
		}
		fmt.Println()
	}
	fmt.Println("===================================")
}
